src/Serilog.Sinks.Map/Sinks/Map/MappedSink`1.cs

```diff
@@ -24,41 +24,86 @@ class MappedSink<TKey> : ILogEventSink, IDisposable
     {
         readonly Func<LogEvent, TKey> _keySelector;
         readonly Action<TKey, LoggerSinkConfiguration> _configure;
+        readonly int? _sinkMapCountLimit;
         readonly object _sync = new object();
         readonly Dictionary<TKey, Logger> _sinkMap = new Dictionary<TKey, Logger>();
+        bool _disposed;
 
-        public MappedSink(Func<LogEvent, TKey> keySelector, Action<TKey, LoggerSinkConfiguration> configure)
+        public MappedSink(Func<LogEvent, TKey> keySelector,
+                          Action<TKey, LoggerSinkConfiguration> configure,
+                          int? sinkMapCountLimit)
         {
             _keySelector = keySelector;
             _configure = configure;
+            _sinkMapCountLimit = sinkMapCountLimit;
         }
 
         public void Emit(LogEvent logEvent)
         {
             var key = _keySelector(logEvent);
 
-            Logger sink;
             lock (_sync)
             {
-                if (!_sinkMap.TryGetValue(key, out sink))
+                if (_disposed)
+                    throw new ObjectDisposedException(nameof(MappedSink<TKey>), "The mapped sink has been disposed.");
+
+                if (_sinkMap.TryGetValue(key, out var existing))
                 {
-                    var config = new LoggerConfiguration()
-                        .MinimumLevel.Is(LevelAlias.Minimum);
+                    existing.Write(logEvent);
+                    return;
+                }
+
+                var config = new LoggerConfiguration()
+                    .MinimumLevel.Is(LevelAlias.Minimum);
+
+                _configure(key, config.WriteTo);
+                var sink = config.CreateLogger();
 
-                    _configure(key, config.WriteTo);
-                    sink = _sinkMap[key] = config.CreateLogger();
+                if (_sinkMapCountLimit == 0)
+                {
+                    using (sink)
+                        sink.Write(logEvent);
                 }
-            }
+                else if (_sinkMapCountLimit == null || _sinkMapCountLimit > _sinkMap.Count)
+                {
+                    // This case is a little faster as no EH nor iteration is required
+                    _sinkMap[key] = sink;
+                    sink.Write(logEvent);
+                }
+                else
+                {
+                    _sinkMap[key] = sink;
+                    try
+                    {
+                        sink.Write(logEvent);
+                    }
+                    finally
+                    {
+                        while(_sinkMap.Count > _sinkMapCountLimit.Value)
+                        {
+                            foreach (var k in _sinkMap.Keys)
```

Comment:
Just noting here our earlier discussion.. this means sinks are released from the map in a somewhat random way, and potentially a future enhancement could be to make the eviction more predictable (i.e. LIFO). This feature still does make things generally better, because being able to limit the count allows us to stop unlimited resources from being consumed.